crates/iceberg/src/expr/predicate.rs

```diff
@@ -195,6 +214,94 @@ impl UnboundPredicate {
     pub fn or(self, other: UnboundPredicate) -> UnboundPredicate {
         UnboundPredicate::Or(LogicalExpression::new([Box::new(self), Box::new(other)]))
     }
+
+    /// Returns a predicate representing the negation ('NOT') of this one,
+    /// by using inverse predicates rather than wrapping in a `NOT`.
+    /// Used for `NOT` elimination.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use std::ops::Bound::Unbounded;
+    /// use iceberg::expr::BoundPredicate::Unary;
+    /// use iceberg::expr::{LogicalExpression, UnboundPredicate, UnboundReference};
+    /// use iceberg::spec::Datum;
+    /// let expr1 = UnboundReference::new("a").less_than(Datum::long(10));
+    /// let expr2 = UnboundReference::new("b").less_than(Datum::long(5)).and(UnboundReference::new("c").less_than(Datum::long(10)));
+    ///
+    /// let result = expr1.negate();
+    /// assert_eq!(&format!("{result}"), "a >= 10");
+    ///
+    /// let result = expr2.negate();
+    /// assert_eq!(&format!("{result}"), "(b >= 5) OR (c >= 10)");
+    /// ```
+    pub fn negate(self) -> UnboundPredicate {
+        match self {
+            UnboundPredicate::And(expr) => UnboundPredicate::Or(LogicalExpression::new(
+                expr.inputs.map(|expr| Box::new(expr.negate())),
+            )),
+            UnboundPredicate::Or(expr) => UnboundPredicate::And(LogicalExpression::new(
+                expr.inputs.map(|expr| Box::new(expr.negate())),
+            )),
+            UnboundPredicate::Not(expr) => {
+                let LogicalExpression { inputs: [input_0] } = expr;
+                input_0.negate()
+            }
+            UnboundPredicate::Unary(expr) => UnboundPredicate::Unary(match expr.op {
+                PredicateOperator::IsNull => {
+                    UnaryExpression::new(PredicateOperator::NotNull, expr.term)
+                }
+                PredicateOperator::NotNull => {
+                    UnaryExpression::new(PredicateOperator::IsNull, expr.term)
```

Comment:
I would suggest add an `negate` method for predicate operator, and then we can simplify these with following:
```
UnaryExpression::new(op.negate(), expr.term)
```

Same for `Binary` and `Set`